### PR TITLE
Overhaul user API, role flow

### LINKED ISSUE #119: Move over UserAPI to Supabase

### PR DESCRIPTION
**Pull request by SekoiaTree**
Fixes #119. Currently incomplete.

This PR moves the User API over to Supabase, and additionally overhauls the role system, as well as applying to an association and such.

### CODE REVIEW

File: app/src/main/java/com/github/se/assocify/model/CurrentUser.kt [hunk 1]
```diff
@@ -1,6 +1,9 @@
 package com.github.se.assocify.model
 
+import com.github.se.assocify.model.entities.User
+
 object CurrentUser {
   var userUid: String? = null
   var associationUid: String? = null
+  var user: User? = null
```
Comment (Mai-LinhC) [inline app/src/main/java/com/github/se/assocify/model/CurrentUser.kt:8]: we have to keep either uid or user -> discussed in group
probably need to make another PR for that to merge before this one : this one is already too big to do it here
if not, we at least need to put it in scrum board as task for (if possible) this sprint

File: app/src/main/java/com/github/se/assocify/model/entities/Role.kt [hunk 2]
```diff
@@ -22,22 +22,14 @@ data class Role(val name: String) {
...
     return when (name.lowercase()) {
-      "pending" -> RoleType.PENDING
-      "presidence" -> RoleType.PRESIDENCE
+      "presidency" -> RoleType.PRESIDENCY
       "treasury" -> RoleType.TREASURY
       "committee" -> RoleType.COMMITTEE
       "member" -> RoleType.MEMBER
       "staff" -> RoleType.STAFF
-      else -> RoleType.PENDING
+      else -> RoleType.STAFF
```
Comment (Mai-LinhC) [inline app/src/main/java/com/github/se/assocify/model/entities/Role.kt:30]: we need to keep a pending role : used for when the user is not yet accepted in asso / when you create the asso and the role is not set yet

File: app/src/main/java/com/github/se/assocify/model/entities/Role.kt [hunk 3]
```diff
@@ -22,22 +22,14 @@ data class Role(val name: String) {
...
-   * Whether the role is an active role
-   *
-   * @return whether the role is an active role
-   */
-  fun isAnActiveRole(): Boolean {
-    return getRoleType() != RoleType.PENDING
-  }
 }
+
+data class PermissionRole(val uid: String, val associationId: String, val type: RoleType)
```
Comment (Mai-LinhC) [inline app/src/main/java/com/github/se/assocify/model/entities/Role.kt:35]: add a comment on what it is??

File: app/src/main/java/com/github/se/assocify/model/entities/Role.kt [hunk 4]
```diff
@@ -1,19 +1,19 @@
...
+  MEMBER,
+  STAFF
+}
+
 data class Role(val name: String) {
   constructor() : this("")
 
   /*
    * enumeration of all roles, sorted by importance
    */
```
Comment (Mai-LinhC) [inline app/src/main/java/com/github/se/assocify/model/entities/Role.kt:16]: move comment :')

File: app/src/main/java/com/github/se/assocify/model/entities/User.kt [hunk 5]
```diff
@@ -1,17 +1,15 @@
...
-
-  fun toggleRole(role: String): User {
-    return if (hasRole(role)) {
-      User(uid, name, Role("pending"))
-    } else {
-      User(uid, name, Role(role))
-    }
-  }
+@Serializable
+data class User(val uid: String, val name: String, val email: String? = null) {
```
Comment (Mai-LinhC) [inline app/src/main/java/com/github/se/assocify/model/entities/User.kt:6]: why add the email??

File: app/src/main/java/com/github/se/assocify/ui/screens/createAssociation/CreateAssociationScreen.kt [hunk 6]
```diff
@@ -171,19 +171,17 @@ fun CreateAssociationScreen(
...
-                                },
-                                modifier = Modifier.testTag("roleitem-${role.name}"))
-                          }
+                      RoleType.entries.forEach { role ->
+                        ListItem(
+                            headlineContent = { Text(role.name.uppercase()) },
+                            trailingContent = {
+                              RadioButton(
+                                  modifier = Modifier.testTag("role-${role.name}"),
+                                  selected = false, // state.editMember!!.hasRole(role.name),
```
Comment (Mai-LinhC) [inline app/src/main/java/com/github/se/assocify/ui/screens/createAssociation/CreateAssociationScreen.kt:180]: need to know what role people have -> see other comments

File: app/src/main/java/com/github/se/assocify/model/entities/User.kt [hunk 7]
```diff
@@ -1,17 +1,15 @@
...
+  constructor() : this("", "")
 }
```
Comment (Mai-LinhC) [inline app/src/main/java/com/github/se/assocify/model/entities/User.kt:8]: since you removed the toggle role, there still should be somewhere to (dis)able a role for someone in an association : doesn't need to be done here but needs to be put as a task to do in scrum board, otherwise we might forget

File: app/src/main/java/com/github/se/assocify/model/entities/User.kt [hunk 8]
```diff
@@ -1,17 +1,15 @@
...
+
+data class AssociationMember(
+    val uid: String,
+    val associationId: String,
+    val name: String,
```
Comment (Mai-LinhC) [inline app/src/main/java/com/github/se/assocify/model/entities/User.kt:13]: I'm not sure i get what uid/name represent, if not the uid/name of the User ? if so, why not just put a user instead ??

File: app/src/main/java/com/github/se/assocify/ui/screens/createAssociation/CreateAssociationViewmodel.kt [hunk 9]
```diff
@@ -32,7 +32,7 @@ class CreateAssociationViewmodel(
    * Utility function : Sorts the members by role then name
    */
   private fun sortMembers(memberList: List<User>): List<User> {
-    return memberList.sortedWith(compareBy({ it.role.getRoleType().ordinal }, { it.name }))
+    return memberList.sortedWith(compareBy({ it.name }))
```
Comment (Mai-LinhC) [inline app/src/main/java/com/github/se/assocify/ui/screens/createAssociation/CreateAssociationViewmodel.kt:35]: you can't just remove stuff :''''))))))
it's supposed to be first sorted by role (because the roles have different importance), then by name
i get that this PR was already too big and the roles were messy but at least leave the previous state of stuff in comment and add a note somewhere (scrum board...) to do this again once we solve the other problems concerning roles etc

File: app/src/main/java/com/github/se/assocify/ui/screens/createAssociation/CreateAssociationViewmodel.kt
Comment (Mai-LinhC) [inline app/src/main/java/com/github/se/assocify/ui/screens/createAssociation/CreateAssociationViewmodel.kt:1]: note for later PR : need to redo once roles are clean

**Review by Mai-LinhC — COMMENTED** [review]
notes when testing on emulator : cant change my name in DB 

will finish reviewing later, i'm at 13/26 files.........
please do smaller PRs as we said </3

File: app/src/main/java/com/github/se/assocify/model/entities/Role.kt (hunk 2 above)
Comment (SekoiaTree) [reply app/src/main/java/com/github/se/assocify/model/entities/Role.kt:30]: I disagree. Roles are only given to members of the association. When a user doesn't have a role yet, it can be `null` temporarily.

File: app/src/main/java/com/github/se/assocify/model/entities/User.kt (hunk 5 above)
Comment (SekoiaTree) [reply app/src/main/java/com/github/se/assocify/model/entities/User.kt:6]: This is the actual data stored in the database. An email is useful for contacting a user outside of the app.

File: app/src/main/java/com/github/se/assocify/model/entities/User.kt (hunk 8 above)
Comment (SekoiaTree) [reply app/src/main/java/com/github/se/assocify/model/entities/User.kt:13]: An AssociationMember also has a role. I don't know why I added the name, though.

File: app/src/main/java/com/github/se/assocify/model/entities/User.kt (hunk 7 above)
Comment (SekoiaTree) [reply app/src/main/java/com/github/se/assocify/model/entities/User.kt:8]: This is dealt with by AssociationMember.

**Review by Mai-LinhC — COMMENTED** [review]
Looks pretty good, I need to test on device then it'll be done !

**Comment by Mai-LinhC** [comment]
I can't save receipt anymore (with an account that was already logged in) :(

**Review by Mai-LinhC — APPROVED** [review]
Looks all good now, thanks :)
